go/api/config.go

```diff
@@ -0,0 +1,309 @@
+// given password.
+func NewRedisCredentialsWithDefaultUsername(password string) *RedisCredentials {
+	return &RedisCredentials{password: password}
+}
+
+func (creds *RedisCredentials) toProtobuf() *protobuf.AuthenticationInfo {
+	return &protobuf.AuthenticationInfo{Username: creds.username, Password: creds.password}
+}
+
+// ReadFrom represents the client's read from strategy.
+type ReadFrom int
+
+const (
+	// Primary - Always get from primary, in order to get the freshest data.
+	Primary ReadFrom = iota
+	// PreferReplica - Spread the requests between all replicas in a round-robin manner. If no replica is available, route the
+	// requests to the primary.
+	PreferReplica
+)
+
+func mapReadFrom(readFrom ReadFrom) protobuf.ReadFrom {
+	if readFrom == PreferReplica {
+		return protobuf.ReadFrom_PreferReplica
+	}
+
+	return protobuf.ReadFrom_Primary
+}
+
+type baseClientConfiguration struct {
+	addresses      []NodeAddress
+	useTLS         bool
+	credentials    *RedisCredentials
+	readFrom       ReadFrom
+	requestTimeout int
+	clientName     string
+}
+
+func (config *baseClientConfiguration) toProtobuf() *protobuf.ConnectionRequest {
+	request := protobuf.ConnectionRequest{}
+	for _, address := range config.addresses {
+		request.Addresses = append(request.Addresses, address.toProtobuf())
+	}
+
+	if config.useTLS {
+		request.TlsMode = protobuf.TlsMode_SecureTls
+	} else {
+		request.TlsMode = protobuf.TlsMode_NoTls
+	}
+
+	if config.credentials != nil {
+		request.AuthenticationInfo = config.credentials.toProtobuf()
+	}
+
+	request.ReadFrom = mapReadFrom(config.readFrom)
+	if config.requestTimeout != 0 {
+		request.RequestTimeout = uint32(config.requestTimeout)
+	}
+
+	if config.clientName != "" {
+		request.ClientName = config.clientName
+	}
+
+	return &request
+}
+
+// BackoffStrategy represents the strategy used to determine how and when to reconnect, in case of connection failures. The
+// time between attempts grows exponentially, to the formula:
+//
+//	rand(0 ... factor * (exponentBase ^ N))
+//
+// where N is the number of failed attempts.
+//
+// Once the maximum value is reached, that will remain the time between retry attempts until a reconnect attempt is successful.
+// The client will attempt to reconnect indefinitely.
+type BackoffStrategy struct {
+	// Required: number of retry attempts that the client should perform when disconnected from the server, where the time
+	// between retries increases. Once the retries have reached the maximum value, the time between retries will remain
+	// constant until a reconnect attempt is successful.
+	numOfRetries int
+	// Required: the multiplier that will be applied to the waiting time between each retry.
+	factor int
+	// Required: the exponent base configured for the strategy.
+	exponentBase int
+}
+
+// NewBackoffStrategy returns a new BackoffStrategy with the given configuration parameters.
+func NewBackoffStrategy(numOfRetries int, factor int, exponentBase int) *BackoffStrategy {
```

Comment:
why are you returning by pointer and not by value?

Reply:
We could do either. [This post](https://stackoverflow.com/a/23551970/22506474) suggests using pointers for big structs and structs you need to change, and values otherwise. But it also says that when pointers/values perform similarly the Go-y approach is to choose whatever provides the more natural semantics. Since the RedisClientConfiguration is a large-ish struct I made its constructor return a pointer. This struct is smaller, but I returned a pointer to be consistent with the RedisClientConfiguration constructor. Let me know if you'd prefer returning the value for small structs.

Reply:
As long as we're consistent, either is fine.